ibis/backends/base/__init__.py

```diff
@@ -206,7 +210,82 @@ def _ipython_key_completions_(self) -> list[str]:
         return self._backend.list_tables()
 
 
-class BaseBackend(abc.ABC):
+# should have a better name
+class ResultHandler:
+    @staticmethod
+    def _import_pyarrow():
+        try:
+            import pyarrow
+        except ImportError:
+            raise ModuleNotFoundError(
+                "Exporting to arrow formats requires `pyarrow` but it is not installed"  # noqa: ignore
+            )
+        else:
+            return pyarrow
+
+    def to_pylist(
+        self,
+        expr: ir.Expr,
+        *,
+        params: Mapping[ir.Scalar, Any] | None = None,
+        limit: int | str | None = None,
+        **kwargs: Any,
+    ) -> list[tuple[Any, ...]]:
+        return toolz.concat(
+            self.to_pylist_batches(expr, params=params, limit=limit, **kwargs)
+        )
+
+    def to_pylist_batches(
+        self,
+        expr: ir.Expr,
+        *,
+        params: Mapping[ir.Scalar, Any] | None = None,
+        limit: int | str | None = None,
+        chunk_size: int = 1_000_000,
+        **kwargs: Any,
+    ) -> Iterable[list]:
+        raise NotImplementedError
+
+    def to_pyarrow(
+        self,
+        expr: ir.Expr,
+        *,
+        params: Mapping[ir.Scalar, Any] | None = None,
+        limit: int | str | None = None,
+        **kwargs: Any,
+    ) -> pa.Table:
+        pa = self._import_pyarrow()
+        try:
+            return pa.Table.from_batches(
+                self.to_pyarrow_batches(
+                    expr, params=params, limit=limit, **kwargs
+                )
+            )
+        except ValueError:
+            # The pyarrow batches iterator is empty so pass in an empty
+            # iterator and the pyarrow schema
+            return pa.Table.from_batches([], schema=expr.schema().to_pyarrow())
+
+    def to_pyarrow_batches(
+        self,
+        expr: ir.Expr,
+        *,
+        params: Mapping[ir.Scalar, Any] | None = None,
+        limit: int | str | None = None,
+        chunk_size: int = 1_000_000,
+        **kwargs: Any,
+    ) -> Iterable[pa.RecordBatch]:
```

Comment:
IMO it would be preferable to return a `RecordBatchReader`, which allows introspecting the schema without needing get any of the batches out of the iterable.

Reply:
Cool, I can definitely swap that in.

Question: do you know if I pass in an iterator of batches to `RecordBatchReader.from_batches` are they consumed eagerly or will it defer until the `read_next_batch` call?

Reply:
It will defer.